src/coreclr/jit/importer.cpp

```diff
@@ -860,6 +860,19 @@ void Compiler::impPopCallArgs(CORINFO_SIG_INFO* sig, GenTreeCall* call)
         {
             // Morph trees that aren't already OBJs or MKREFANY to be OBJs
             assert(ti.IsType(TI_STRUCT));
+
+            // The argument and parameter types can be different for legitimate
+            // reasons, but we expect them to be compatible in those cases. One
+            // example where this happens is when inlining shared code into a
+            // non-generic function, in which case we might see the __Canon in
+            // the parameter type but exact types in the signature type.
+            //
+            // TODO-ARGS: Remove this quirk; we should be able to use the
```

Comment:
I probably shouldn't call it a quirk, but we currently store the argument's class handle instead of the signature's class handle in `CallArg::m_signatureClsHnd`. They are different in some cases, e.g. when inlining shared code where the inlining function has exact types. This todo is about storing the actual signature type (returned by the EE) instead of the argument's type. It's probably one of the first follow-ups I'll have after this PR so I'll avoid churning CR to clarify (unless there's other feedback).